MAINT: More cleanup of einsum

jaimefrio opened the pull request:
Somewhat simplifies the code inside get_single_op_view, moves logic out of the main einsum function loop, and fixes #11221.

jaimefrio committed:
MAINT: Move call to get_single_op_view out of loop.

jaimefrio committed:
MAINT: Refactor get_combined_dims_view.

Moves check for repeated labels in operands out of the main einsum loop
and into get_combined_dims_view. The code there is slightly simplified.

Also, fixes #11221 by only rewriting the labels array once all labels
have been processed.

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2103,96 +2102,108 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
+         * The char type may be either signed or unsigned, we
+         * need it to be signed here.
+         */
+        int label  = (signed char)labels[idim];
```
nit: extra space here

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2103,96 +2102,108 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
 }
 
+static int
+_no_labels_are_negative(char *labels, int ndim)
```
Maybe just pass a `signed char*` instead?

jaimefrio committed:
MAINT: Make _no_labels_are_negative parameter signed char.

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2092,83 +2091,101 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
+        /*
+         * The char type may be either signed or unsigned, we
+         * need it to be signed here.
+         */
```
This comment might make more sense above the function definition

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2092,83 +2091,101 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
+         */
+        int label = labels[idim];
+
+        if (label < 0) {
```
Could just write `if (labels[idim] < 0)` here, now that there's no casting to worry about

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2092,83 +2091,101 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
 }
 
+static int
+_no_labels_are_negative(signed char *labels, int ndim)
```
I'd maybe prefer `_any_labels_are_negative`, and then use it with a preceding `!`. Functions with inverted names are a little confusing - also `any` is common python terminology

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2092,83 +2091,101 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
     npy_intp new_dims[NPY_MAXDIMS];
-    int idim, ndim, icombine, combineoffset;
+    int idim, icombine;
     int icombinemap[NPY_MAXDIMS];
```
Any chance you could comment `icombinemap`, assuming you've worked out what it does?

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2092,83 +2091,101 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
+        npy_intp dim = PyArray_DIM(op, idim);
+        npy_intp stride = PyArray_STRIDE(op, idim);
+
+        /* A repeated label, find the original one and merge them. */
```
This would make more sense to me with the `if` and `else` swapped - since that way the write to `icombinemap` is seen before the read from it.

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2092,83 +2091,101 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
-                            PyArray_DESCR(op),
-                            ndim, new_dims, new_strides,
+                            Py_TYPE(op), PyArray_DESCR(op),
+                            icombine, new_dims, new_strides,
```
This will conflict with #11246 (but not in a difficult way)

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2092,83 +2091,101 @@ get_single_op_view(PyArrayObject *op, int  iop, char *labels,
-            icombinemap[idim] = icombine;
-        }
-        /* If the label is 0, it's an unlabeled broadcast dimension */
-        if (label == 0) {
```
Why did this case go away?

jaimefrio replied:
Because icombinemap is set to zero, it didn't really do anything different than for any other non-negative label.

charris commented:
Needs a rebase.

jaimefrio committed:
MAINT: address review comments.

jaimefrio commented:
PTAL. I think I have addressed all of Eric's comments, which included the need for a rebase.

mattip commented:
Tests now fail, somehting went wrong with the rebase?

jaimefrio committed:
MAINT: Fix small bug after rebasing.

The view was being created with the same number of dimensions as
the original operand, instead of the number of unique labels.

jaimefrio commented:
Yes, the combined dimension view was being created with the same number of dimensions as the original array, instead of the proper value, using uninitialised memory for the sizes and strides of the extra dimensions. That is now fixed, lets see if Travis agrees...

mattip commented:
CI passes, any additional comments?

mattip commented:
@jaimefrio ping, waiting for your reaction

mattip commented:
@jaimefrio, @eric-wieser can this go in as-is?

mattip commented:
@jaimefrio, @eric-wieser can this go in as-is?

eric-wieser commented:
Yep, no behavior change here. Will take one more pass of review, but probably is already fine

mattip commented:
Thanks @jaimefrio and reviewers
